src/java/org/apache/cassandra/service/reads/ReadCallback.java

```diff
@@ -56,29 +59,104 @@
-            maxTombstoneAbortsCount.accumulateAndGet(tombstones, Math::max);
+            aborts.incrementAndGet();
+            maxAbortsCount.accumulateAndGet(value, Math::max);
+        }
+    }
+
+    private interface ToString
+    {
+        String apply(int count, long value, String cql);
+    }
+
+    private class WarningCounter
+    {
+        final WarnAbortCounter tombstones = new WarnAbortCounter();
+        final WarnAbortCounter localReadSizeTooLarge = new WarnAbortCounter();
+        final WarnAbortCounter rowIndexTooLarge = new WarnAbortCounter();
+
+        private String cql, loggableTokens;
+        private String cql()
+        {
+            if (cql == null)
+                cql = command.toCQLString();
+            return cql;
+        }
+        private String loggableTokens()
+        {
+            if (loggableTokens == null)
+                loggableTokens = command.loggableTokens();
+            return loggableTokens;
+        }
+
+        private void trackAborts(WarnAbortCounter counter, TableMetrics.TableMeter metric, ToString toString)
+        {
+            if (counter.aborts.get() > 0)
+            {
+                String msg = toString.apply(counter.aborts.get(), counter.maxAbortsCount.get(), cql());
+                ClientWarn.instance.warn(msg + " with " + loggableTokens());
+                logger.warn(msg);
+                metric.mark();
+            }
+        }
+
+        private void trackWarnings(WarnAbortCounter counter, TableMetrics.TableMeter metric, ToString toString)
+        {
+            if (counter.warnings.get() > 0)
+            {
+                String msg = toString.apply(counter.warnings.get(), counter.maxWarningCount.get(), cql());
+                ClientWarn.instance.warn(msg + " with " + loggableTokens());
+                logger.warn(msg);
+                metric.mark();
+            }
+        }
+
+        void track()
+        {
+            trackAborts(  tombstones, cfs().metric.clientTombstoneAborts,   ReadCallback::tombstoneAbortMessage);
+            trackWarnings(tombstones, cfs().metric.clientTombstoneWarnings, ReadCallback::tombstoneWarnMessage);
+
+            trackAborts(  localReadSizeTooLarge, cfs().metric.clientLocalReadSizeTooLargeAborts,   ReadCallback::localReadSizeTooLargeAbortMessage);
+            trackWarnings(localReadSizeTooLarge, cfs().metric.clientLocalReadSizeTooLargeWarnings, ReadCallback::localReadSizeTooLargeWarnMessage);
+
+            trackAborts(rowIndexTooLarge, cfs().metric.rowIndexSizeTooLargeAborts, ReadCallback::rowIndexTooLargeAbortMessage);
+            trackWarnings(rowIndexTooLarge, cfs().metric.rowIndexSizeTooLargeWarnings, ReadCallback::rowIndexTooLargeWarnMessage);
+        }
+
+        void mayAbort(int received)
+        {
+            if (tombstones.aborts.get() > 0)
+                throw new TombstoneAbortException(tombstones.aborts.get(), Math.toIntExact(tombstones.maxAbortsCount.get()), cql(), resolver.isDataPresent(),
```

Comment:
```suggestion
                throw new TombstoneAbortException(tombstones.aborts.get(), tombstones.maxAbortsCount.get(), cql(), resolver.isDataPresent(),
```
...assuming we change the argument type.